src/daft-plan/src/optimization/rules/push_down_limit.rs

```diff
@@ -0,0 +1,226 @@
+use std::sync::Arc;
+
+use common_error::DaftResult;
+
+use crate::{source_info::SourceInfo, LogicalPlan};
+
+use super::{ApplyOrder, OptimizerRule, Transformed};
+
+/// Optimization rules for pushing Limits further into the logical plan.
+#[derive(Default)]
+pub struct PushDownLimit {}
+
+impl PushDownLimit {
+    pub fn new() -> Self {
+        Self {}
+    }
+}
+
+impl OptimizerRule for PushDownLimit {
+    fn apply_order(&self) -> ApplyOrder {
+        ApplyOrder::TopDown
+    }
+
+    fn try_optimize(&self, plan: Arc<LogicalPlan>) -> DaftResult<Transformed<Arc<LogicalPlan>>> {
+        let limit = match plan.as_ref() {
+            LogicalPlan::Limit(limit) => limit,
+            _ => return Ok(Transformed::No(plan)),
+        };
+        let child_plan = limit.input.as_ref();
+        let new_plan = match child_plan {
+            LogicalPlan::Repartition(_) | LogicalPlan::Coalesce(_) | LogicalPlan::Project(_) => {
+                // Naive commuting with unary ops.
+                //
+                // Limit-UnaryOp -> UnaryOp-Limit
+                let new_limit = plan.with_new_children(&[child_plan.children()[0].clone()]);
```

Comment:
I was thinking about this as well! For this particular case, we should probably align `LogicalPlan::children()` and `LogicalPlan::with_children()` such that this could be reduced to
```rust
let new_limit = plan.with_new_children(child_plan.children()[..]);
```
where the assertion on `child_plan.children()` returning the expected number of children happens in the `plan.with_new_children()` validation. We may have to change `with_new_children` to take `&[&Arc<LogicalPlan>]` or something like that.

As for the more general problem, I think the latter would be the best for the sake of reuse! Having `child_plan.child()?` would be nice.